src/game_logic/turn.ts

```diff
@@ -87,32 +89,50 @@ export function resetWorkers (state: GameState): GameState {
 export function resolveContracts (initialState: GameState): GameState {
   let state = initialState
   // TODO: fix staleness
-  for (const contract of state.player.contracts.open) {
-    state = contractMaybeComplete(state, contract)
-    if (contract.completed) {
-      state = contractCollectRewards(state, contract)
-    }
+  let contractIndex: number | null
+
+  contractIndex = findIndexOfSatisfiedOpenContract(state)
+  while (contractIndex !== null && contractIndex >= 0) {
+    state = completeContract(state, contractIndex)
+    contractIndex = findIndexOfSatisfiedOpenContract(state)
   }
 
-  state = organizeContracts(state)
-  return state
+  return openNewContracts(state)
 }
 
-function contractMaybeComplete (state: GameState, contract: Contract): GameState {
-  throw new Error('Not implemented')
+/**
+ * Returns the index of the first open contract whose conditions have
+ * been satisfied or, if none was found, null.
+ */
+function findIndexOfSatisfiedOpenContract (state: GameState): number | null {
+  return state.player.contracts.open.findIndex(contract => contract.isSatisfied(state))
```

Comment:
findIndex returns -1 instead of null, so you can update the return type and comment here, and remove the null check up there.